### ISSUE TITLE
roll up updates into single PR

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#4404]
There two main reasons why people group updates today:
- Because updates will fail unless grouped together (e.g. monorepos)
- To reduce the number of PRs and noise associated

The downside with the second one is that if one update breaks things, then the entire PR is failing and none can easily be updated.

I propose a new concept - `rollupName` - that would replace the concept of `groupName` for the majority of users of the second case above. In short, it would "roll up" separate updates into combined PRs only once they passed tests in their own branch. The rolled up PR hopefully then also passes.

`rollupName` would essentially sit "above" `groupName`. e.g. you would use `groupName` for packages that *must* be updated together, while `rollupName` can be used to combine multiple groups if desired. For example you'd still keep monorepos using `groupName` but the "group all" would be better suited for a `rollupName`.

The most common scenario is fairly simple and obvious - a single rollup PR with a handful of updates, all of which passed their individual tests in branches. The challenge comes with the edge cases, e.g.
- If you had an update to v1.0.1 in the rollup PR and then v1.0.2 is released, what do you do? e.g. keep v1.0.1 in the PR but replace it with v1.0.2 if it passes?
- If you need to rebase an "individual" update branch because of conflict or staleness, what do you do with the rollup PR it was already in? e.g. do you remove it until you've retested?
- Do you keep it even though it may not now pass tests? And if you remove it and it was the last update in a PR, do you close that PR and reopen it later once you have passing updates again?
- How do you combine it with scheduling, e.g. "give me a rollup PR every Sunday, then stop updating it"? Users might potentially want to have different schedules for when branches are created/updated (e.g. because of CI) versus when rollup PRs are created/updated.

**Comment by grit96** [comment]
Another thing to consider is if automerge is enabled, could the individual branches be merged if they pass before the rollup branch has passed?

**Comment by rarkins** [comment]
> Another thing to consider is if automerge is enabled, could the individual branches be merged if they pass before the rollup branch has passed?

No, the idea of the individual branches is for them to be branch-only for the purpose of testing only and not merged unless a developer chose to merge them or create a PR manually (which shouldn't typically happen).

**Comment by jlsjonas** [comment]
My 2 cents on the edge cases:

> If you had an update to v1.0.1 in the rollup PR and then v1.0.2 is released, what do you do? e.g. keep v1.0.1 in the PR but replace it with v1.0.2 if it passes?

I would avoid frequent changes to the roll-up pr as that would also mean tests would have to re-run.
So in this case I'd indeed keep 1.0.1 in the pr as-is, and thus just update the pr when a new individual branch passes (as proposed by the author).

Generally, the roll-up should try to be updated in batches (i.e. when several new updates are being processed, the roll-up shouldn't be updated until there's no pending builds anymore)
Regarding builds: we'll need a parameter to define the amount of builds required to pass (as this affects merge-ability)

> If you need to rebase an "individual" update branch because of conflict or staleness, what do you do with the rollup PR it was already in? e.g. do you remove it until you've retested?

I would wait to update until the tests show a pass or fail.

> Do you keep it even though it may not now pass tests? And if you remove it and it was the last update in a PR, do you close that PR and reopen it later once you have passing updates again?

Yes.

> How do you combine it with scheduling, e.g. "give me a rollup PR every Sunday, then stop updating it"? Users might potentially want to have different schedules for when branches are created/updated (e.g. because of CI) versus when rollup PRs are created/updated.

This sounds like something that is already in place; rebases still happen outside of scheduling hours but that's it

**Comment by jlsjonas** [comment]
> No, the idea of the individual branches is for them to be branch-only for the purpose of testing only and not merged unless a developer chose to merge them or create a PR manually (which shouldn't typically happen).

I would open a PR for failed branches to give them visibility.

**Comment by rarkins** [comment]
Here's how I think this can work, with a configuration option `rollupName`.

For each base branch:
 - After all upgrade branches are progressed, check them for `rollupName` entries
 - If a rollup exists but all branches are pending, then:
   * If the PR already exists, don't modify the branch but instead update the body to say that it is awaiting pending branches
   * If no PR exists, do nothing (do not create it)
 - If any branches in the rollup have passed:
   * Ensure branch/PR

We can also introduce a new `prCreation` value called `on-error` which means PRs are only created if in error.. otherwise they're suppressed and can be put into a rollup. We can change the default value of `prCreation` to be `auto` which will then default to `immediate` if not a rollup or `on-error` if a rollup.

**Comment by rarkins** [comment]
Answering my own edge cases:

> * If you had an update to v1.0.1 in the rollup PR and then v1.0.2 is released, what do you do? e.g. keep v1.0.1 in the PR but replace it with v1.0.2 if it passes?

If v1.0.1 was the only thing in the rollup PR then the PR would be marked as pending.

If other updates are in the rollup PR then v1.0.1 would be temporarily removed while v1.0.2 is under test.

> * If you need to rebase an "individual" update branch because of conflict or staleness, what do you do with the rollup PR it was already in? e.g. do you remove it until you've retested?

Same as above - reduce the PR if some updates remain as passing, or mark it as pending/WIP if all need rebasing.

> * Do you keep it even though it may not now pass tests? And if you remove it and it was the last update in a PR, do you close that PR and reopen it later once you have passing updates again?

If the latest version fails tests, then it no longer qualifies for the rollup PR.

> * How do you combine it with scheduling, e.g. "give me a rollup PR every Sunday, then stop updating it"? Users might potentially want to have different schedules for when branches are created/updated (e.g. because of CI) versus when rollup PRs are created/updated.

We can allow a separate schedule for rollup PRs.

**Comment by aminya** [comment]
_Copying my comment originally posted in https://github.com/renovatebot/renovate/issues/1342#issuecomment-798823378_

I was able to configure renovate the way I wanted using the following `renovate.json`. 

Now, I only get two PRs every weekend for dependencies and devDependencies. devDependencies are merged automatically if the CI passes, and dependencies require manual merging:

```json
{
  "schedule": ["every weekend"],
  "labels": ["dependencies"],
  "separateMajorMinor": "false",
  "packageRules": [
    {
      "matchDepTypes": ["devDependencies"],
      "matchUpdateTypes": ["major", "minor", "patch", "pin", "digest", "lockFileMaintenance", "rollback", "bump"],
      "groupName": "devDependencies",
      "semanticCommitType": "chore",
      "automerge": true
    },
    {
      "matchDepTypes": ["dependencies"],
      "matchUpdateTypes": ["major", "minor", "patch", "pin", "digest", "lockFileMaintenance", "rollback", "bump"],
      "groupName": "dependencies",
      "semanticCommitType": "fix"
    }
  ]
}
```

**Comment by laurentsimon** [comment]
I'm curious if there's any activity or ETA for this feature. I think this is a fantastic feature and it would reduce noise to the extent of making hash pinning/lockfile dependency update close to painless.

**Comment by regexident** [comment]
+1 on what @laurentsimon said above.

I'd like to add another scenario where roll-up is kind of a make it or break it factor:

### Updating multiple interdependent dependencies

One of my projects [depends on a group of 10+ dependencies](https://github.com/regexident/cargo-modules/blob/main/Cargo.toml#L27-L38) that get updated on a weekly basis and in lock-step, and for which a PR would only succeed if it updated all of them at once, rather than creating individual PRs for each dependency.

My hope was to lessen the burden of manually updating to the latest versions every week by use of Renovate, alas in its current form it would [create about 10+ individual PRs](https://github.com/regexident/cargo-modules/pull/83) every week, which then would all fail.

**Comment by rarkins** [comment]
> ### Updating multiple interdependent dependencies
> One of my projects [depends on a group of 10+ dependencies](https://github.com/regexident/cargo-modules/blob/main/Cargo.toml?rgh-link-date=2022-06-29T11%3A55%3A40Z#L27-L38) that get updated on a weekly basis and in lock-step, and for which a PR would only succeed if it updated all of them at once, rather than creating individual PRs for each dependency.

This is actually not a use case for rollups, because rollups would be based on the individual branches *passing* first. So if your dependencies need updating in lock step, they would likely fail individually and never be part of the rollup branch. So they would need *grouping* first before they could then be rolled up with other dependencies.

I don't want to get distracted off rollups too far, but for that use case:
- [today] You can always define your own grouping using packageRules
- [today] These packages could/should be added to our built-in grouping rules so that everyone can benefit
- [future] We will change our default branch naming to be based off source URL instead of dependency name, so that all packages from the same URL will by default be grouped together without needing config

**Comment by regexident** [comment]
Oh neat! Thanks for the explanation. Looks like I'll have to look into grouping then. Noted.

**Comment by algomaster99** [comment]
Hey! I wanted to try this feature out using the GitHub market place. I see one closed PR related to this - https://github.com/renovatebot/renovate/pull/10592/files. Has there been some progress, in a separate PR maybe?

**Comment by felipesere** [comment]
We are interested in this rollup feature too. Any way we can support?

**Comment by viceice** [comment]
no, it's very complicated to implement.

**Comment by ced455** [comment]
Hello, 
I am also interested by this feature so we can run pipeline on week end 
https://github.com/renovatebot/renovate/pull/10592  as been closed but the code was ok according to the comment, do you need anything else ?

**Comment by mmkal** [comment]
For anyone interested, I am trying out a makeshift way of getting something similar. Basically, set the renovate target branch to `deps` instead of `main`, tell it to automerge without a pull request if possible, and then use a GitHub Action to keep `deps` and `main` in sync where possible. I'm trying it on an OSS repo to see how it works out.

Some more [explanation on Twitter here](https://twitter.com/mmkalmmkal/status/1806690658456862998) and here's the [GitHub Action which syncs between `deps` and `main`](https://github.com/mmkal/trpc-cli/blob/main/.github/workflows/deps.yml).

@rarkins @viceice would be interested your thoughts on whether this is a good/bad idea!

**Comment by HonkingGoose** [comment]
## Splitting off dependencies?

How do we want to handle this case?

1. Renovate creates a roll up PR.
2. The user wants to split off a specific dependency from the rolled up group.

How about giving the user checkboxes to split off a dependency into a new single PR?

### Mockup

#### List of rolled up dependencies

If you want to split off a dependency into a separate PR, select the checkbox for that dependency:

- [ ] `rolled-up-dependency-1`
- [ ] `rolled-up-dependency-2`
- [ ] `rolled-up-dependency-3`

### Code for mockup

```markdown
#### List of rolled up dependencies

If you want to split off a dependency into a separate PR, select the checkbox for that dependency:

- [ ] `rolled-up-dependency-1`
- [ ] `rolled-up-dependency-2`
- [ ] `rolled-up-dependency-3`
```

**Comment by mathisgauthey** [comment]
Hi there,

I am very much interested in this feature. Is there any update on that PR ?

Thanks in advance !
